File: model/user_token.go
package model

import "time"

// UserTokenStatus 用户 API Key 状态。
const (
	// UserTokenStatusActive 正常可用。
	UserTokenStatusActive = "active"
	// UserTokenStatusDisabled 用户主动禁用。
	UserTokenStatusDisabled = "disabled"
	// UserTokenStatusExhausted 余额耗尽（Sprint 2 渠道选择器后续在 consume 时自动转）。
	UserTokenStatusExhausted = "exhausted"
	// UserTokenStatusExpired 已过期（CurrentAuthUserByToken 命中 ExpiredAt 时自动转）。
	UserTokenStatusExpired = "expired"
)

// UserToken 用户自建 API Key（OpenAI 兼容 sk- 格式）。
//
// 用途：
//  1) 外部 OpenAI SDK / Cursor / Cline / curl 直接对接 Freedom 当网关用
//  2) Admin 端按 token 维度审计用量、限速、封禁
//
// 字段约束：
//   - KeyHash：sha256(明文) hex；唯一索引；明文永不落库，创建时一次性返回
//   - KeyPrefix：明文前 12 位（含 sk-fk-），用于前端列表识别
//   - ModelLimits / AllowIPs：白名单，空字符串 = 不限
//   - BalanceCapCents > 0 时只扣本字段额度；否则扣用户全局余额
//   - UnlimitedBalance = true 时 BalanceCapCents 字段被忽略
type UserToken struct {
	ID               string     `json:"id" gorm:"primaryKey"`
	UserID           string     `json:"userId" gorm:"index"`
	Name             string     `json:"name"`
	KeyPrefix        string     `json:"keyPrefix" gorm:"type:varchar(32)"`
	KeyHash          string     `json:"-" gorm:"type:varchar(128);uniqueIndex"`
	Status           string     `json:"status" gorm:"type:varchar(16);index"`
	ExpiredAt        *time.Time `json:"expiredAt,omitempty"`
	BalanceCapCents  int        `json:"balanceCapCents"`
	UsedCents        int        `json:"usedCents"`
	UnlimitedBalance bool       `json:"unlimitedBalance"`
	ModelLimits      string     `json:"modelLimits" gorm:"type:varchar(512)"`
	AllowIPs         string     `json:"allowIps" gorm:"type:varchar(512)"`
	LastUsedIP       string     `json:"lastUsedIp"`
	LastUsedAt       *time.Time `json:"lastUsedAt,omitempty"`
	CreatedAt        time.Time  `json:"createdAt"`
	UpdatedAt        time.Time  `json:"updatedAt"`
}

// MaskUserTokenKey 给前端展示的打码 key：前 7 + "..." + 末 4（new-api MaskTokenKey 风格）。
// 输入应是 KeyPrefix 或明文；输出始终是脱敏后的展示串。
func MaskUserTokenKey(key string) string {
	if len(key) <= 11 {
		return key
	}
	return key[:7] + "..." + key[len(key)-4:]
}

File: web/src/services/api/user_token.ts
import { apiDelete, apiGet, apiPost } from "@/services/api/request";

export type UserTokenStatus = "active" | "disabled" | "exhausted" | "expired";

// 单条 user_token 记录（与后端 model/user_token.go 对齐）。
// KeyHash 字段 GORM tag json:"-" 屏蔽，永远不会出现在前端。
export type UserToken = {
    id: string;
    userId: string;
    name: string;
    // 后端已脱敏："sk-fk-...1234" 格式
    keyPrefix: string;
    status: UserTokenStatus;
    // null = 永不过期
    expiredAt: string | null;
    // 0 = 用用户全局余额
    balanceCapCents: number;
    usedCents: number;
    unlimitedBalance: boolean;
    // "" = 不限
    modelLimits: string;
    allowIps: string;
    lastUsedIp: string;
    lastUsedAt: string | null;
    createdAt: string;
    updatedAt: string;
};

export type UserTokenCreateRequest = {
    name: string;
    expiredAt?: string;
    balanceCapCents?: number;
    unlimitedBalance?: boolean;
    // modelLimits / allowIps 后端已支持，但 Sprint 1.5 UI 不暴露配置入口
    modelLimits?: string[];
    allowIps?: string[];
};

export type UserTokenCreateResponse = {
    token: UserToken;
    // 明文 sk-fk-...：仅在创建时一次性返回，关闭弹窗后无法再看到
    raw: string;
};

export type UserTokenListResponse = {
    items: UserToken[];
    total: number;
};

export async function listUserTokens(token: string) {
    return apiGet<UserTokenListResponse>("/api/v1/user-tokens", undefined, token);
}

export async function createUserToken(token: string, req: UserTokenCreateRequest) {
    return apiPost<UserTokenCreateResponse>("/api/v1/user-tokens", req, token);
}

export async function deleteUserToken(token: string, id: string) {
    return apiDelete<{ code: number; data: { ok: boolean }; msg: string }>(
        `/api/v1/user-tokens/${id}`,
        token,
    );
}

export async function disableUserToken(token: string, id: string) {
    return apiPost<{ code: number; data: { ok: boolean }; msg: string }>(
        `/api/v1/user-tokens/${id}/disable`,
        {},
        token,
    );
}

export async function enableUserToken(token: string, id: string) {
    return apiPost<{ code: number; data: { ok: boolean }; msg: string }>(
        `/api/v1/user-tokens/${id}/enable`,
        {},
        token,
    );
}
